codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -69,61 +69,96 @@
-     */
-    public Http2ConnectionHandler(Http2ConnectionDecoder decoder,
-                                  Http2ConnectionEncoder encoder) {
-        this.initialSettings = null;
-        this.decoder = checkNotNull(decoder, "decoder");
-        this.encoder = checkNotNull(encoder, "encoder");
-        if (encoder.connection() != decoder.connection()) {
-            throw new IllegalArgumentException("Encoder and Decoder do not share the same connection object");
+        public final Http2Settings initialSettings() {
+            return initialSettings;
         }
-    }
 
-    public Http2ConnectionHandler(Http2Connection connection, Http2Settings initialSettings) {
-        this(connection, new DefaultHttp2FrameReader(), new DefaultHttp2FrameWriter(),
-                initialSettings);
-    }
+        @SuppressWarnings("unchecked")
+        public B server(boolean isServer) {
+            server = isServer;
+            return (B) this;
+        }
+
+        @SuppressWarnings("unchecked")
+        public B frameLogger(Http2FrameLogger logger) {
+            frameLogger = logger;
+            return (B) this;
+        }
+
+        @SuppressWarnings("unchecked")
+        public B gracefulShutdownTimeoutMillis(long gracefulShutdownTimeoutMillis) {
+            this.gracefulShutdownTimeoutMillis = gracefulShutdownTimeoutMillis;
+            return (B) this;
+        }
 
-    public Http2ConnectionHandler(Http2Connection connection, Http2FrameReader frameReader,
-                                  Http2FrameWriter frameWriter, Http2Settings initialSettings) {
-        this.initialSettings = initialSettings;
-        encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
-        decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, frameReader);
+        public final T build() {
+            return build(new DefaultHttp2Connection(server));
+        }
+
+        public final T build(Http2Connection connection) {
+            Http2FrameReader reader = new DefaultHttp2FrameReader(validateHttp2Headers);
+            Http2FrameWriter writer = new DefaultHttp2FrameWriter();
+            if (frameLogger != null) {
+                reader = new Http2InboundFrameLogger(reader, frameLogger);
+                writer = new Http2OutboundFrameLogger(writer, frameLogger);
+            }
+            Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, writer);
+            Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, reader);
+            return build(decoder, encoder);
+        }
+
+        public final T build(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder) {
+            // Call the abstract build method
+            T handler = build0(decoder, encoder);
+
+            // Setup post build options
+            handler.gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
+            handler.decoder().frameListener(frameListener);
```

Comment:
Yeah, I'm fine with adding the accessor back in.  An alternative would be to throw if we try to set the listener on the decoder twice.  Then this code would just be:

```java
if (frameListener != null) {
    decoder.frameListener(frameListener);
}
```